### PR TITLE
Next hop forwarder

### PR DESCRIPTION
**Pull request by dpino**
Import of Marcel Wiget's `nh_fwd` [1] and `next_hop app` [2].

Next hop forwarder is an app that classifies traffic. It's one of the fundamental pieces of SnabbVMX [3] Depending on nh_fwd's configuration parameters, traffic is redirected to a service link or to a VM link. Likewise, incoming traffic to a service link is redirected to a VM link or wire link, and finally incoming VM traffic is redirected to the wire or a service.

[1] https://github.com/mwiget/snabb/tree/1to1_mapping/src/apps/nh_fwd
[2] https://github.com/mwiget/snabb/tree/1to1_mapping/src/program/snabbvmx/nexthop
[3] http://www.ch-open.ch/fileadmin/OCD_2016/3_Marcel_Wiget.pdf

### CODE REVIEW

**Comment by dpino** [comment]
As for the test with a VM which I originally planed to do, that test was finally deployed in the Snabbvmx PR https://github.com/Igalia/snabb/pull/391. Snabbvmx relies on `nh_fwd`, so when Snabbvmx is tested with a VM, `nh_fwd` is being tested too.

As a single app, the unit tests added are to `nh_fwd` are sufficient to validate the app routes traffic to the correct links (wire, service, vm) depending on type of packet.

**Comment by kbara** [comment]
```
% luacheck nh_fwd.lua | grep unused  
    nh_fwd.lua:19:14: unused variable ntohs
    nh_fwd.lua:20:7: unused variable htonl
    nh_fwd.lua:20:14: unused variable ntohl
    nh_fwd.lua:23:7: unused variable bor
    nh_fwd.lua:23:12: unused variable lshift
    nh_fwd.lua:31:7: unused variable n_ethertype_ipv6
    nh_fwd.lua:122:22: unused argument self
    nh_fwd.lua:231:22: unused argument self
```

**Comment by kbara** [comment]
`nh_fwd.lua:111:10: variable ipv4_checksum is never accessed`

File: src/program/lwaftr/nexthop/nexthop.lua [hunk 1]
```diff
@@ -0,0 +1,68 @@
+module(..., package.seeall)
+
+local S = require("syscall")
+local ethernet = require("lib.protocol.ethernet")
+local lib = require("core.lib")
+local shm = require("core.shm")
+
+local ffi = require("ffi")
+local C = ffi.C
```
Comment (kbara) [inline src/program/lwaftr/nexthop/nexthop.lua]: C is unused, remove this line.

File: src/program/lwaftr/nexthop/nexthop.lua [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   function handlers.h (arg) usage(0) end
+   return lib.dogetopt(args, handlers, "h", long_opts)
+end
+
+local function is_current_process (pid)
+   return pid == S.getpid()
+end
+
+function run (args)
+   args = parse_args(args)
```
Comment (kbara) [inline src/program/lwaftr/nexthop/nexthop.lua]: args is never used, just call parse_args(args) here for the side effects, or assign it to _.

File: src/program/lwaftr/nexthop/nexthop.lua [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+local ethernet = require("lib.protocol.ethernet")
+local lib = require("core.lib")
+local shm = require("core.shm")
+
+local ffi = require("ffi")
+local C = ffi.C
+
+local macaddress_t = ffi.typeof[[
+struct { uint8_t ether[6]; }
+]]
```
Comment (kbara) [inline src/program/lwaftr/nexthop/nexthop.lua:11]: Note for the future (doesn't need to be addressed before merging): we have a bunch of ctype pollution with various 6-byte ethernet wrappers.

**Comment by kbara** [comment]
Kudos to Marcel - this is nice code. I'm happy to LGTM it once a few nits with things like unused variables are fixed, Diego.

**Comment by dpino** [comment]
I removed all unused variables. As for `nh_fwd.lua:111:10: variable ipv4_checksum is never accessed`, it was a false positive. `ipv4_checksum` is a cdata pointer. It's written several times but it's never read, that's fine.

**Pull request merged**